src/main/java/net/dv8tion/jda/bot/sharding/DefaultShardManager.java

```diff
@@ -0,0 +1,577 @@
+
+            if (api == null)
+                api = this.buildInstance(shardId);
+            else if (api.getStatus() == JDA.Status.RECONNECT_QUEUED)
+                ((JDAImpl) api).getClient().reconnect(true, true);
+
+            // as this happens before removing the shardId if from the queue just try again after 5 seconds
+        }
+        catch (RateLimitedException e)
+        {
+            LOG.warn("Hit the login ratelimit while creating new JDA instances");
+        }
+        catch (LoginException e)
+        {
+            // this can only happen if the token has been changed
+            // in this case the ShardManager will just shutdown itself as there currently is no way of hot-swapping the token on a running JDA instance.
+            LOG.warn("The token has been invalidated and the ShardManager will shutdown!", e);
+            this.shutdown();
+        }
+        catch (final Exception e)
+        {
+            LOG.error("Caught an exception in the queue processing thread", e);
+        }
+
+        this.shards.getMap().put(shardId, api);
+        this.queue.remove(shardId);
+    }
+
+    protected JDAImpl buildInstance(final int shardId) throws LoginException, RateLimitedException
+    {
+        final JDAImpl jda = new JDAImpl(AccountType.BOT, this.token, this.httpClientBuilder, this.wsFactory, this.shardedRateLimiter,
+            this.autoReconnect, this.enableVoice, this.enableBulkDeleteSplitting, this.enableBulkDeleteSplitting, retryOnTimeout,
+            this.corePoolSize, this.maxReconnectDelay);
+
+        jda.asBot().setShardManager(this);
+
+        if (this.eventManager != null)
+            jda.setEventManager(this.eventManager);
+
+        if (this.audioSendFactory != null)
+            jda.setAudioSendFactory(this.audioSendFactory);
+
+        this.listeners.forEach(jda::addEventListener);
+        jda.setStatus(JDA.Status.INITIALIZED); //This is already set by JDA internally, but this is to make sure the listeners catch it.
+
+        // Set the presence information before connecting to have the correct information ready when sending IDENTIFY
+        PresenceImpl presence = ((PresenceImpl) jda.getPresence());
+        if (gameProvider != null)
+            presence.setCacheGame(this.gameProvider.apply(shardId));
+        if (idleProvider != null)
+            presence.setCacheIdle(this.idleProvider.apply(shardId));
+        if (statusProvider != null)
+            presence.setCacheStatus(this.statusProvider.apply(shardId));
+
+        if (this.gatewayURL == null)
+        {
+            Pair<String, Integer> gateway = jda.getGatewayBot().complete();
+
+            this.gatewayURL = gateway.getLeft();
+
+            if (this.shardsTotal == -1)
+            {
+                this.shardsTotal = gateway.getRight();
+
+                this.shards = new ShardCacheViewImpl(this.shardsTotal);
+
+                for (int i = 0; i < shardsTotal; i++)
```

Comment:
You have too many empty lines in this section that just stretch this code for no apparent reason